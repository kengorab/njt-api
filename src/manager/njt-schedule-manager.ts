import * as NJTScheduleDao from '../dao/njt-schedule-dao'
import * as moment from 'moment'
import globals from '../globals'
import { ScheduleResult } from '../types'

function parseTrainInfo(trainInfo: string): [string, number] {
  const [trainLine, trainNumber] = trainInfo.split(' #')
  return [trainLine, parseInt(trainNumber)]
}

function formatTime(time: string): string {
  if (time[0] === '0') {
    return time.substring(1)
  }
  return time
}

export async function getScheduleFromNJTPage(
  originStation: string,
  originId: string,
  destStation: string,
  destinationId: string,
  date: moment.Moment
): Promise<ScheduleResult[]> {
  const njtPageText = await NJTScheduleDao.getScheduleFromNJTPage(
    originStation,
    originId,
    destStation,
    destinationId,
    date
  )

  const { cheerio } = globals
  const $ = cheerio.load(njtPageText)
  const table = $('b:contains(Origin)')
    .parents('table')
    .first()
  return table
    .find('tr')
    .toArray()
    .filter((_, idx) => idx !== 0)
    .map(row => {
      const $tr = $(row)
      const $tds = $tr.children('td')

      let $origin, $transfer, $destination, $travelTime
      if ($tds.length === 4) {
        ;[$origin, $transfer, $destination, $travelTime] = $tds.toArray()
      } else {
        ;[$origin, $destination, $travelTime] = $tds.toArray()
      }

      // Origin data
      const $originSpan = $($origin)
        .children('span')
        .first()
      const [originTime, originTrainInfo] = $originSpan
        .contents()
        .toArray()
        .map(n => $(n).text())
        .filter(text => text !== '')
      const [originLine, originTrainNumber] = parseTrainInfo(originTrainInfo)
      const origin = {
        time: formatTime(originTime),
        trainLine: originLine,
        trainNumber: originTrainNumber
      }

      // Transfer data (if applicable)
      let transfer
      if ($transfer) {
        const $transferSpan = $($transfer)
          .children('span')
          .first()
        if ($transferSpan.children().length !== 0) {
          const [arriveTime, station, departTime, trainInfo] = $transferSpan
            .contents()
            .toArray()
            .map(n => $(n).text())
            .filter(text => text !== '')

          const arrivalTime = arriveTime.replace('Arrive ', '')
          const departureTime = departTime.replace('Depart ', '')

          const waitTime = moment(departureTime, 'HH:mm A').diff(
            moment(arrivalTime, 'HH:mm A'),
            'minutes'
          )

          const [trainLine, trainNumber] = parseTrainInfo(trainInfo)
          transfer = {
            arrivalTime: formatTime(arrivalTime),
            station,
            departureTime: formatTime(departureTime),
            trainLine,
            trainNumber,
            waitTime
          }
        }
      }

      // Destination data
      const arrivalTime = $($destination)
        .children('span')
        .first()
        .text()

      // Travel time data
      const travelTime = $($travelTime)
        .children('span')
        .first()
        .text()
        .split(' ')[0]

      return {
        origin,
        transfer,
        arrivalTime: formatTime(arrivalTime),
        travelTime: parseInt(travelTime)
      }
    })
}
