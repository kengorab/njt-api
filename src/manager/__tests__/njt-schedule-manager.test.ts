jest.mock('../../dao/njt-schedule-dao')
jest.mock('../../globals')

import * as NJTScheduleManager from '../njt-schedule-manager'
import * as moment from 'moment'

describe('getScheduleFromNJTPage', () => {
  it('should return schedule results (with no transfers)', async () => {
    const schedule = await NJTScheduleManager.getScheduleFromNJTPage(
      'Millburn',
      'some_id',
      'New York Penn Station',
      'some_id',
      moment()
    )
    expect(schedule).toMatchSnapshot()
  })

  it('should return schedule results (with transfers)', async () => {
    const schedule = await NJTScheduleManager.getScheduleFromNJTPage(
      'Aberdeen Matawan',
      'some_id',
      'Allenhurst',
      'some_id',
      moment()
    )
    expect(schedule).toMatchSnapshot()
  })
})
