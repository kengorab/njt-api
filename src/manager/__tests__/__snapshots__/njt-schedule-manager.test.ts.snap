// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getScheduleFromNJTPage should return schedule results (with no transfers) 1`] = `
Array [
  Object {
    "arrivalTime": "6:45 AM",
    "origin": Object {
      "time": "5:50 AM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6908,
    },
    "transfer": undefined,
    "travelTime": 55,
  },
  Object {
    "arrivalTime": "7:39 AM",
    "origin": Object {
      "time": "6:50 AM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6910,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "8:39 AM",
    "origin": Object {
      "time": "7:50 AM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6912,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "9:39 AM",
    "origin": Object {
      "time": "8:50 AM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6914,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "10:39 AM",
    "origin": Object {
      "time": "9:50 AM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6916,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "11:39 AM",
    "origin": Object {
      "time": "10:50 AM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6918,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "12:40 PM",
    "origin": Object {
      "time": "11:50 AM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6920,
    },
    "transfer": undefined,
    "travelTime": 50,
  },
  Object {
    "arrivalTime": "1:39 PM",
    "origin": Object {
      "time": "12:50 PM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6922,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "2:39 PM",
    "origin": Object {
      "time": "1:50 PM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6924,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "3:39 PM",
    "origin": Object {
      "time": "2:50 PM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6926,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "4:39 PM",
    "origin": Object {
      "time": "3:50 PM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6928,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "5:41 PM",
    "origin": Object {
      "time": "4:50 PM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6930,
    },
    "transfer": undefined,
    "travelTime": 51,
  },
  Object {
    "arrivalTime": "6:39 PM",
    "origin": Object {
      "time": "5:50 PM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6932,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "7:39 PM",
    "origin": Object {
      "time": "6:50 PM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6934,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "8:40 PM",
    "origin": Object {
      "time": "7:50 PM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6936,
    },
    "transfer": undefined,
    "travelTime": 50,
  },
  Object {
    "arrivalTime": "9:39 PM",
    "origin": Object {
      "time": "8:50 PM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6938,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "10:39 PM",
    "origin": Object {
      "time": "9:50 PM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6940,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "11:39 PM",
    "origin": Object {
      "time": "10:50 PM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6942,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
  Object {
    "arrivalTime": "12:39 AM",
    "origin": Object {
      "time": "11:50 PM",
      "trainLine": "Morris & Essex",
      "trainNumber": 6944,
    },
    "transfer": undefined,
    "travelTime": 49,
  },
]
`;

exports[`getScheduleFromNJTPage should return schedule results (with transfers) 1`] = `
Array [
  Object {
    "arrivalTime": "7:32 AM",
    "origin": Object {
      "time": "6:40 AM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3215,
    },
    "transfer": Object {
      "arrivalTime": "7:11 AM",
      "departureTime": "7:24 AM",
      "station": "Long Branch",
      "trainLine": "North Jersey Coast",
      "trainNumber": 4315,
      "waitTime": 13,
    },
    "travelTime": 52,
  },
  Object {
    "arrivalTime": "8:53 AM",
    "origin": Object {
      "time": "8:08 AM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3221,
    },
    "transfer": Object {
      "arrivalTime": "8:38 AM",
      "departureTime": "8:45 AM",
      "station": "Long Branch",
      "trainLine": "North Jersey Coast",
      "trainNumber": 4321,
      "waitTime": 7,
    },
    "travelTime": 45,
  },
  Object {
    "arrivalTime": "9:34 AM",
    "origin": Object {
      "time": "8:49 AM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3223,
    },
    "transfer": Object {
      "arrivalTime": "9:23 AM",
      "departureTime": "9:26 AM",
      "station": "Long Branch",
      "trainLine": "North Jersey Coast",
      "trainNumber": 4323,
      "waitTime": 3,
    },
    "travelTime": 45,
  },
  Object {
    "arrivalTime": "11:36 AM",
    "origin": Object {
      "time": "10:53 AM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3231,
    },
    "transfer": Object {
      "arrivalTime": "11:25 AM",
      "departureTime": "11:28 AM",
      "station": "Long Branch",
      "trainLine": "North Jersey Coast",
      "trainNumber": 4331,
      "waitTime": 3,
    },
    "travelTime": 43,
  },
  Object {
    "arrivalTime": "12:34 PM",
    "origin": Object {
      "time": "11:48 AM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3235,
    },
    "transfer": Object {
      "arrivalTime": "12:23 PM",
      "departureTime": "12:26 PM",
      "station": "Long Branch",
      "trainLine": "North Jersey Coast",
      "trainNumber": 4335,
      "waitTime": 3,
    },
    "travelTime": 46,
  },
  Object {
    "arrivalTime": "2:34 PM",
    "origin": Object {
      "time": "1:50 PM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3243,
    },
    "transfer": Object {
      "arrivalTime": "2:23 PM",
      "departureTime": "2:26 PM",
      "station": "Long Branch",
      "trainLine": "North Jersey Coast",
      "trainNumber": 4343,
      "waitTime": 3,
    },
    "travelTime": 44,
  },
  Object {
    "arrivalTime": "3:50 PM",
    "origin": Object {
      "time": "3:13 PM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 2303,
    },
    "transfer": undefined,
    "travelTime": 37,
  },
  Object {
    "arrivalTime": "5:22 PM",
    "origin": Object {
      "time": "4:37 PM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3255,
    },
    "transfer": Object {
      "arrivalTime": "5:07 PM",
      "departureTime": "5:14 PM",
      "station": "Long Branch",
      "trainLine": "North Jersey Coast",
      "trainNumber": 4355,
      "waitTime": 7,
    },
    "travelTime": 45,
  },
  Object {
    "arrivalTime": "6:12 PM",
    "origin": Object {
      "time": "5:36 PM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3361,
    },
    "transfer": undefined,
    "travelTime": 36,
  },
  Object {
    "arrivalTime": "6:39 PM",
    "origin": Object {
      "time": "6:02 PM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3363,
    },
    "transfer": undefined,
    "travelTime": 37,
  },
  Object {
    "arrivalTime": "7:18 PM",
    "origin": Object {
      "time": "6:30 PM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3269,
    },
    "transfer": Object {
      "arrivalTime": "7:04 PM",
      "departureTime": "7:09 PM",
      "station": "Long Branch",
      "trainLine": "North Jersey Coast",
      "trainNumber": 4369,
      "waitTime": 5,
    },
    "travelTime": 48,
  },
  Object {
    "arrivalTime": "7:39 PM",
    "origin": Object {
      "time": "7:01 PM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3373,
    },
    "transfer": undefined,
    "travelTime": 38,
  },
  Object {
    "arrivalTime": "7:54 PM",
    "origin": Object {
      "time": "7:18 PM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 2313,
    },
    "transfer": undefined,
    "travelTime": 36,
  },
  Object {
    "arrivalTime": "8:51 PM",
    "origin": Object {
      "time": "8:10 PM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3279,
    },
    "transfer": Object {
      "arrivalTime": "8:39 PM",
      "departureTime": "8:43 PM",
      "station": "Long Branch",
      "trainLine": "North Jersey Coast",
      "trainNumber": 4379,
      "waitTime": 4,
    },
    "travelTime": 41,
  },
  Object {
    "arrivalTime": "10:08 PM",
    "origin": Object {
      "time": "9:25 PM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3283,
    },
    "transfer": Object {
      "arrivalTime": "9:57 PM",
      "departureTime": "10:00 PM",
      "station": "Long Branch",
      "trainLine": "North Jersey Coast",
      "trainNumber": 4383,
      "waitTime": 3,
    },
    "travelTime": 43,
  },
  Object {
    "arrivalTime": "12:16 AM",
    "origin": Object {
      "time": "11:29 PM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3293,
    },
    "transfer": Object {
      "arrivalTime": "12:05 AM",
      "departureTime": "12:08 AM",
      "station": "Long Branch",
      "trainLine": "North Jersey Coast",
      "trainNumber": 4393,
      "waitTime": 3,
    },
    "travelTime": 47,
  },
  Object {
    "arrivalTime": "1:11 AM",
    "origin": Object {
      "time": "12:23 AM",
      "trainLine": "North Jersey Coast",
      "trainNumber": 3297,
    },
    "transfer": Object {
      "arrivalTime": "12:59 AM",
      "departureTime": "1:03 AM",
      "station": "Long Branch",
      "trainLine": "North Jersey Coast",
      "trainNumber": 4397,
      "waitTime": 4,
    },
    "travelTime": 48,
  },
]
`;
